griptape/drivers/prompt/google_prompt_driver.py

```diff
@@ -37,112 +47,198 @@ class GooglePromptDriver(BasePromptDriver):
+
         inputs = [
-            {"role": self.__to_role(message), "parts": self.__to_content(message)}
+            ContentDict(
+                {
+                    "role": self.__to_google_role(message),
+                    "parts": [self.__to_google_message_content(content) for content in message.content],
+                }
+            )
             for message in prompt_stack.messages
             if not message.is_system()
         ]
 
-        # Gemini does not have the notion of a system message, so we insert it as part of the first message in the history.
-        system_messages = prompt_stack.system_messages
-        if system_messages:
-            inputs[0]["parts"].insert(0, system_messages[0].to_text())
-
         return inputs
 
-    def __prompt_stack_content_message_content(self, content: BaseMessageContent) -> ContentDict | str:
+    def __to_google_role(self, message: Message) -> str:
+        if message.is_assistant():
+            return "model"
+        else:
+            return "user"
+
+    def __to_google_tools(self, tools: list[BaseTool]) -> list[dict]:
+        FunctionDeclaration = import_optional_dependency("google.generativeai.types").FunctionDeclaration
+
+        tool_declarations = []
+        for tool in tools:
+            for activity in tool.activities():
+                schema = (tool.activity_schema(activity) or Schema({})).json_schema("Parameters Schema")
+
+                if "values" in schema["properties"]:
+                    schema = schema["properties"]["values"]
+
+                schema = remove_key_in_dict_recursively(schema, "additionalProperties")
+                tool_declaration = FunctionDeclaration(
+                    name=f"{tool.name}_{tool.activity_name(activity)}",
+                    description=tool.activity_description(activity),
+                    parameters={
+                        "type": schema["type"],
+                        "properties": schema["properties"],
+                        "required": schema.get("required", []),
+                    },
+                )
+
+                tool_declarations.append(tool_declaration)
+
+        return tool_declarations
+
+    def __to_google_message_content(self, content: BaseMessageContent) -> ContentDict | Part | str:
         ContentDict = import_optional_dependency("google.generativeai.types").ContentDict
+        protos = import_optional_dependency("google.generativeai.protos")
 
         if isinstance(content, TextMessageContent):
             return content.artifact.to_text()
         elif isinstance(content, ImageMessageContent):
             return ContentDict(mime_type=content.artifact.mime_type, data=content.artifact.value)
+        elif isinstance(content, ActionCallMessageContent):
+            action = content.artifact.value
+
+            return protos.Part(function_call=protos.FunctionCall(name=action.tag, args=action.input))
```

Comment:
As I understand it the tool call (`function_call`) should only ever be part of messages with the role "model" / assistant.  Is there anything validating / guaranteeing this?

(Similar comment for the other relevant prompt drivers)

Reply:
Yeah we don't really have any client-side validation on the Prompt Drivers to ensure the payload is formatted correctly. While I agree that this would be nice to have, it feels out of scope for this PR.